internal/rpc/proxy_http.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2024 Canonical.
+package rpc
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/juju/zaputil"
+	"github.com/juju/zaputil/zapctx"
+
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+)
+
+type HTTPClientOptions struct {
+	TLSConfig *tls.Config
+	URL       url.URL
+}
+
+func ProxyHTTP(ctx context.Context, ctl *dbmodel.Controller, w http.ResponseWriter, req *http.Request) {
+	var tlsConfig *tls.Config
+	if ctl.CACertificate != "" {
+		cp := x509.NewCertPool()
+		ok := cp.AppendCertsFromPEM([]byte(ctl.CACertificate))
+		if !ok {
+			zapctx.Warn(ctx, "no CA certificates added")
+		}
+		tlsConfig = &tls.Config{
+			RootCAs:    cp,
+			ServerName: ctl.TLSHostname,
+			MinVersion: tls.VersionTLS12,
+		}
+	}
+
+	if ctl.PublicAddress != "" {
+		err := proxyHTTP(ctx, w, req, HTTPClientOptions{
+			TLSConfig: tlsConfig,
+			URL:       httpUrl(*req.URL, ctl.PublicAddress),
+		})
+		if err == nil {
+			return
+		}
+	}
+	for _, hps := range ctl.Addresses {
+		for _, hp := range hps {
```

Comment:
I'm wondering if we shouldn't leverage some of the logic that we already have in the dialer for this.

Reply:
I thought about that, but to me it seems like they behave differently enough that try to use the same function make it more confusing